ykrt/src/trace/swt/patch.rs

```diff
@@ -0,0 +1,79 @@
+use libc::{mprotect, size_t, sysconf, PROT_EXEC, PROT_READ, PROT_WRITE};
+use std::mem;
+use std::{ffi::c_void, sync::Once};
+
+#[cfg(tracer_swt)]
+use crate::yk_trace_basicblock;
+
+#[cfg(tracer_swt)]
+static ORIGINAL_INSTRUCTIONS_INIT: Once = Once::new();
+#[cfg(tracer_swt)]
+static mut ORIGINAL_INSTRUCTIONS: [u8; 1] = [0; 1];
+#[cfg(tracer_swt)]
+static mut PATCH_INSTRUCTIONS: [u8; 1] = [0xC3];
+
+#[cfg(tracer_swt)]
+unsafe fn save_original_instructions(
+    function_ptr: usize,
+    instructions: *mut u8,
+    num_of_instructions: usize,
+) {
+    let func_ptr: *const () = function_ptr as *const ();
+    std::ptr::copy_nonoverlapping(func_ptr as *const u8, instructions, num_of_instructions);
+}
+
+#[cfg(tracer_swt)]
+unsafe fn patch_function(function_ptr: usize, code: *const u8, size: size_t) {
+    let page_size = sysconf(libc::_SC_PAGESIZE) as usize;
+
+    let func_address = ((function_ptr as usize) & !(page_size - 1)) as *mut c_void;
+    let page_size_aligned = (((function_ptr as usize) + mem::size_of_val(&function_ptr))
+        - (func_address as usize)) as usize;
+
+    // Set function memory region to be writable
+    let result = mprotect(
+        func_address,
+        page_size_aligned,
+        PROT_READ | PROT_WRITE | PROT_EXEC,
```

Comment:
You don't need the `PROT_EXEC` here since you are changing it back to executable again below. In fact, on some systems you are not allowed to mark a region writeable and executable at the same time.

Reply:
Updated 👉 3725c7189fb0159b52a412da595f760370cf2ff4

Reply:
Sounds good.
But when I remove `PROT_EXEC` I get some of the tests failing in a non-deterministic way 🤔 
I guess its because somehow test execution overlaps and this runtime patching takes effect while other tests try to execute these instructions.
I get the same result when I run the tests sequentially as:
```shell
YKB_TRACER=swt cargo test -- --test-threads=1
```

Reply:
> In fact, on some systems you are not allowed to mark a region writeable and executable at the same time.

I think Lukas's point is that you can `mprotect(PROT_READ | PROT_WRITE)` or `mprotect(PROT_READ | PROT_EXEC)` but you can't combine `PROT_EXEC` and `PROT_WRITE`: you have to write the page and only then mark it as executable.

Reply:
Updated :point_right: 40c0696e76ca685f7334716461d13816e852ac3e
I get that :)
My issue was with failing the test, but I can't reproduce it anymore :no_mouth:

Reply:
I understand that I can't combine `PROT_WRITE` and `PROT_EXEC` but..

I again get this issue with failing tests if I set `PROT_READ | PROT_WRITE` before calling`copy_nonoverlapping` (patching the function):
```shell
Exited due to signal: 11
```
But when I set it as `PROT_READ | PROT_WRITE | PROT_EXEC` it works fine.

Reply:
The normal trick here is: set to `PROT_READ|PROT_WRITE` do the writes; then set it to `PROT_READ | PROT_EXEC`. That way you never have `PROT_WRITE` *and* `PROT_EXEC` set at once.

Reply:
I understand and I wish it would work for me but I get segfault when I do it.
Will dig deeper.

Reply:
I think I'm going mad :upside_down_face: 
Its working fine with `PROT_READ | PROT_WRITE` on patch and `PROT_READ | PROT_EXEC` restore.
updated :point_right: 6f4bdc3a3a5d4ac15a130d7e3a877a8f9af062c3